proposals/0000-language-editions.rst

```diff
@@ -0,0 +1,1144 @@
+  apparent.
+
+  Although language editions have wide authority, we must be tasteful
+  in how they work. It would be problematic to have ``Stable2024``
+  and ``Stable2027`` disagree on the meaning of widely used features.
+  Yet I think it's best to be maximally expressive here, relying on
+  our future selves not to abuse our power.
+
+  (In the conversation, it was suggested that a ``Python`` language
+  edition would be beyond reason... but actually I think such a thing
+  would be lovely, if the necessary work was done to translate the
+  interfaces between Python and Haskell.)
+
+- For backward compatibility, a language edition can be specified
+  at an arbitrary place in a command-line invocation of GHC, or in a
+  ``LANGUAGE`` pragma in a file. Even though it might come later in
+  a command line, the edition can affect the meaning of command-line
+  arguments that precede it.
+
+- A cabal file will allow a new
+  field ``language-edition``, available both at top-level and in
+  build-product stanzas. This will specify the language edition. To
+  support backward compatibility, this will use the ``default-language``
+  setting if that is available, and omitting the ``language-edition`` will
+  use the default. At some point, it is expected that ``language-edition``
+  will become required.
+
+- Any new language features invented once e.g. ``Stable2027`` is released
+  will not be available with the 2024 editions. That is, if we introduce a new
+  feature ``-XDependentTypes`` in 2028, then enabling ``-XDependentTypes``
+  with ``Stable2024`` (or even ``Experimental2024``) will be an error.
+  This policy encourages users to upgrade their editions in order
+  to access GHC's new features.
+
+- Once an edition has been eclipsed by newer models (that is, once the
+  calendar reads 2028 but someone is still using ``2024``), we know that
+  we are compiling older code. Our priority for such maintenance modes
+  is stability, not, say, making sure that the set of warnings conforms
+  to the latest standard.
+
+- We introduce several coarse-grained *semantic bundles* that group
+  together similar features. The individual choices are detailed below.
+  When using one of the new language
+  editions, error messages suggest opting into one of these bundles,
+  rather than suggesting individual extensions.
+
+  The bundles described below are all just combinations of existing
+  language extensions and warnings. I expect that bundles will
+  remain as such (though it's conceivable that we might imagine
+  language features that do not get their own extension, just a spot
+  in a bundle). It's also conceivable that the bundles will evolve
+  to encompass more expressive power (such as controlling optimization
+  flags or the meaning of ``import Prelude``).
+
+  Unlike language editions, a user may specify any number of bundles
+  when compiling a file. In order to simplify their processing, a
```

Comment:
It might be nice to specify a bundle in a .cabal file yes. Maybe also `-XNoFancyTypes`? I guess `-XNoFancyTypes` means that individual modules would be unable to turn on the features in there? I'll add this as an open question -- could be convinced in any direction here.